### DISCUSSION TITLE
py5.P3D fatal error

### DISCUSSION THREAD

**Discussion by kresdjan** [py5coding/py5generator#539] [Q&A]
Maybe I am using a wrong version of Java Runtime Environment. But if I set the render to `P3D` and run the sketch Java crashes.

```python
import py5


def setup():
    py5.size(300, 300, py5.P3D)    
        

py5.run_sketch()
```
Java Runtime Environment error 
```
python test3d.py 
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGILL (0x4) at pc=0x00007ff819aa37c7, pid=1108, tid=47875
#
# JRE version: OpenJDK Runtime Environment (23.0.1+11) (build 23.0.1+11)
# Java VM: OpenJDK 64-Bit Server VM (23.0.1+11, mixed mode, sharing, tiered, compressed oops, compressed class ptrs, g1 gc, bsd-amd64)
# Problematic frame:
# C  [AppKit+0xe6c7c7]  -[NSWMWindowCoordinator performTransactionUsingBlock:]+0x2fb
#
# No core dump will be written. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# An error report file with more information is saved as:
# /Users/kresdjan/Documents/Code/python/py5/hs_err_pid1108.log
[2.103s][warning][os] Loading hsdis library failed
#
# If you would like to submit a bug report, please visit:
#   https://trac.macports.org/newticket?port=openjdk23
# The crash happened outside the Java Virtual Machine in native code.
# See problematic frame for where to report the bug.
#

Compilation abort trap: 6 at Sat Nov  2 16:56:59
```
Platform information 

```
sw_vers 
ProductName:		macOS
ProductVersion:		15.1
BuildVersion:		24B83
```
```
java --version
openjdk 23.0.1 2024-10-15
OpenJDK Runtime Environment (build 23.0.1+11)
OpenJDK 64-Bit Server VM (build 23.0.1+11, mixed mode, sharing)
```

**Comment by villares** [comment]
Hi @kresdjan, thanks for reporting this. 

Can you give us the *py5* version you are using? 

More recent versions of py5 will warn you if you have a *headless* Java JDK (which won't work). Could that be the case?

**Reply by kresdjan** [reply]
py5 0.10.2a0

**Comment by hx2A** [comment]
Hi @kresdjan ! Your computer is MacOS. Is it one of the Apple silicon machines?

Do you have to use openjdk 23? Can you try openjdk 17 instead?

**Reply by kresdjan** [reply]
No it is an Intel Mac. No I don't have to use openjdk 23, I just installed the latest version. I can try with 17 instead.

**Reply by kresdjan** [reply]
```
python test3d.py 
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGILL (0x4) at pc=0x00007ff80e81e7c7, pid=892, tid=58631
#
# JRE version: OpenJDK Runtime Environment Temurin-17.0.13+11 (17.0.13+11) (build 17.0.13+11)
# Java VM: OpenJDK 64-Bit Server VM Temurin-17.0.13+11 (17.0.13+11, mixed mode, tiered, compressed oops, compressed class ptrs, g1 gc, bsd-amd64)
# Problematic frame:
# C  [AppKit+0xe6c7c7]  -[NSWMWindowCoordinator performTransactionUsingBlock:]+0x2fb
#
# No core dump will be written. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# An error report file with more information is saved as:
# /Users/kresdjan/Documents/Code/python/py5/hs_err_pid892.log
#
# If you would like to submit a bug report, please visit:
#   https://github.com/adoptium/adoptium-support/issues
# The crash happened outside the Java Virtual Machine in native code.
# See problematic frame for where to report the bug.
#
```

**Comment by hx2A** [comment]
How did you install Java?

Can you try installing the Processing PDE and testing a simple 3d sketch?

**Reply by kresdjan** [reply]
I installed Java via MacPorts `sudo port install openjdk17-temurin` https://ports.macports.org/port/openjdk17-temurin/.
3D renders without a problem in the Processing PDE.
<img width="1032" alt="Skærmbillede 2024-11-02 kl  20 43 37" src="https://github.com/user-attachments/assets/15b63fb5-4fec-4d67-bcfe-3c97e70ddfea">

**Reply by hx2A** [reply]
I am not familiar with that method of installing Java.

The Processing PDE comes with its own copy of the JRE. Since Processing works fine on your machine, the problem most likely is because of how py5 is accessing Java.

Can you remove the Java you installed and install it as suggested on the [install page](https://py5coding.org/content/install.html)?

```bash
pip install install-jdk
python -c "import jdk; print('Java installed to', jdk.install('17'))"
```

**Reply by kresdjan** [reply]
Sure. I gave it a go. Uninstalled Java 23 and 17 ports. But it doesn't seem to be installed correctly. `java --version` is unable to locate the runtime, even after I've set `JAVA_HOME` and restarted the computer.

```
$ java --version
The operation couldn’t be completed. Unable to locate a Java Runtime.
Please visit http://www.java.com for information on installing Java.
```
```
$echo $JAVA_HOME
/Users/kresdjan/.jdk/jdk-17.0.13+11
```
When running the sketch produces the same result as before..
```
$ python test3d.py
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGILL (0x4) at pc=0x00007ff81a7537c7, pid=963, tid=51203
#
# JRE version: OpenJDK Runtime Environment Temurin-17.0.13+11 (17.0.13+11) (build 17.0.13+11)
# Java VM: OpenJDK 64-Bit Server VM Temurin-17.0.13+11 (17.0.13+11, mixed mode, tiered, compressed oops, compressed class ptrs, g1 gc, bsd-amd64)
# Problematic frame:
# C  [AppKit+0xe6c7c7]  -[NSWMWindowCoordinator performTransactionUsingBlock:]+0x2fb
#
# No core dump will be written. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# An error report file with more information is saved as:
# /Users/kresdjan/Documents/Code/python/py5/hs_err_pid963.log
#
# If you would like to submit a bug report, please visit:
#   https://github.com/adoptium/adoptium-support/issues
# The crash happened outside the Java Virtual Machine in native code.
# See problematic frame for where to report the bug.
#
```
The JRE version and Java VM seem wrong. It looks like what I had before installing Java with `install-jdk` and it doesn't match the example listed in the docs.  

The `getDefaultJVMPath()` does point to the right java.
```
/Users/kresdjan/.jdk/jdk-17.0.13+11/Contents/Home/lib/libjli.dylib
```

**Reply by hx2A** [reply]
> But it doesn't seem to be installed correctly. java --version is unable to locate the runtime, even after I've set JAVA_HOME and restarted the computer.

That's fine. You'd need `java` to be on your `$PATH` for that to work. py5 doesn't need that when it starts the JVM though. It uses JAVA_HOME, and if that is not set, it will look in a few known locations (including were install-jdk puts it)

As long as this works:

```python
import jpype
jpype.startJVM()
```

py5 will be able to start the JVM and do what it needs to do.

I don't understand why py5 isn't working for you. I am running low on ideas. Can you tell me how you installed python? Are you using anaconda, or some other method? Are you running in a virtual environment of some kind, either with anaconda or something else? If not, can you give this a try? Anaconda might be a good choice here because in addition to python libraries it can install other system libraries that opengl will need to work (which for some reason doesn't work on your machine). Anaconda can ensure you have a clean environment with all the right system libraries and python libraries. Sometimes a machine you've had for a while can accumulate nonsense that causes problems like this.

**Reply by kresdjan** [reply]
@hx2A No sweat Jim, there is properly a million different reasons for it not working for me. You have been a big help already. I just installed python using my preferred package manager MacPorts. That is it really, I don't have Anaconda and I don't use virtual environments, I just installed `py5` for my user into `/Users/kresdjan/Library/Python/3.12`. I can try and give Anaconda a go tomorrow when I have time.

**Reply by Maltiec** [reply]
well, my experience helping one of my students ended up with editing .zshrc file and putting JAVA path there. That was the only way to tell WHERE is Java. )
